### PR TITLE
Overhaul of incentives

### PR DESCRIPTION
**Pull request by 0xcacti**
Adjusted incentive mechanism to work on smaller groups and remove pairwise comparison to allow for regular weight setting. 

## Summary by CodeRabbit

- **New Features**
  - Introduced a persistent database for storing and retrieving miner scores, improving reliability of score tracking.
  - Added robust utility functions for reading and writing competition commitments to the blockchain, with comprehensive error handling and logging.
  - Implemented a new scoring system based on loss, benchmark, and power parameters for fairer and more transparent evaluation.
  - Enhanced model training with configurable LoRA hyperparameters, increased context length, and improved dataset handling for reproducibility.
  - Added support for new model templates and training configurations, including Qwen2.5-1.5B-Instruct and TinyLlama-1.1B-Chat.
  - Provided detailed logging throughout miner and validator workflows for improved traceability and debugging.
  - Added comprehensive unit tests for database operations, scoring logic, and chain commitment utilities.

- **Bug Fixes**
  - Improved error handling and fallback mechanisms when reading chain commitments and during model evaluation and training.
  - Fixed potential issues with duplicate evaluation scores by detecting and zeroing out duplicates.

- **Refactor**
  - Replaced print statements with structured logging for clearer diagnostics.
  - Simplified and streamlined validator logic, removing outdated consensus and win-rate mechanisms.
  - Updated imports and project structure to use the new package namespace.

- **Documentation**
  - Updated README to reflect new project name, model configurations, and improved clarity on usage instructions.

- **Chores**
  - Expanded `.gitignore` and updated dependencies for enhanced development and testing support.
  - Added and updated configuration files for training and package management.

- **Style**
  - Improved code readability and formatting across several modules.

### CODE REVIEW

File: neurons/validator.py [hunk 1]
```diff
@@ -16,204 +16,338 @@
...
+                    bt.logging.info(
+                        f"Downloading evaluation dataset: {constants.eval_namespace}/{constants.eval_commit}"
+                    )
+
+                    download_dataset(
+                        constants.eval_namespace,
+                        constants.eval_commit,
+                        local_dir=eval_data_dir,
+                        cache_dir=self.config.cache_dir,
+                    )
```
[the remark this answers is not in the document]
Comment (nickcom007) [reply neurons/validator.py]: @0xcacti  Is this intentional?

File: FLockDataset/validator/validator_utils.py [hunk 2]
```diff
@@ -1,184 +1,10 @@
...
+        bt.logging.warning("Loss is None, returning score of 0")
+        return 0
 
-    def next(self):
-        if len(self.queue) == 0:
-            self.seed, self.queue = self._get_shuffled()
-            self.epochs += 1
-        return self.seed, self.queue.pop()
+    exp = -loss * constants.DECAY_RATE / benchmark_loss
+    return constants.NUM_UIDS**exp
```
[the remark this answers is not in the document]
Comment (nickcom007) [reply FLockDataset/validator/validator_utils.py:14]: @0xcacti  sounds like a valida concern?

File: FLockDataset/validator/validator_utils.py [hunk 3]
```diff
@@ -1,184 +1,14 @@
...
+        bt.logging.error("Invalid benchmark_loss (%s). Returning score of 0.", benchmark_loss)
+        return 0
 
-    def next(self):
-        if len(self.queue) == 0:
-            self.seed, self.queue = self._get_shuffled()
-            self.epochs += 1
-        return self.seed, self.queue.pop()
+    exp = -loss * constants.DECAY_RATE / benchmark_loss
+    return constants.NUM_UIDS**exp
```
[the remark this answers is not in the document]
Comment (nickcom007) [reply FLockDataset/validator/validator_utils.py:14]: @0xcacti A valid suggestion

File: tests/FlockDataset/utils/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    print(f"Reading data for hotkey: {SUBNET_OWNER_HOTKEY}")
+
+    comp = read_chain_commitment(SUBNET_OWNER_HOTKEY, node, subnet_uid)
+
+    print(f"Read data type: {type(comp)}, value: {comp}")
+
+    assert comp is not None, "Should return a valid commitment"
+    assert comp.id == "42", "ID should be 42"
+    assert comp.bench == 100, "Bench should be 100"
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply tests/FlockDataset/utils/test_utils.py]: Ignoring.  fixed on chain.  Easy to fix if test breaks.

File: neurons/validator.py [hunk 5]
```diff
@@ -16,204 +16,338 @@
...
-from FLockDataset.utils.chain import assert_registered
-from FLockDataset.validator.chain import retrieve_model_metadata, set_weights_with_err_msg
-from FLockDataset.validator.validator_utils import EvalQueue, compute_wins, adjust_for_vtrust
-from FLockDataset.validator.trainer import train_lora, download_dataset, clean_cache_folder
+from FLockDataset.utils.chain import assert_registered, read_chain_commitment
+from FLockDataset.validator.chain import (
+    retrieve_model_metadata,
+    set_weights_with_err_msg,
+)
+from FLockDataset.validator.validator_utils import compute_score
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:31]: Using impl in validator_utils.

File: neurons/validator.py [hunk 6]
```diff
@@ -16,204 +16,338 @@
...
+            help="Directory to store evaluation datasets.",
         )
 
         bt.subtensor.add_args(parser)
         bt.logging.add_args(parser)
         bt.wallet.add_args(parser)
         bt.axon.add_args(parser)
         config = bt.config(parser)
+        bt.logging.info(f"Parsed config: {config}")
         return config
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:86]: Fixed.

File: neurons/validator.py [hunk 7]
```diff
@@ -16,204 +16,338 @@
...
+        bt.logging.info(f"Weights initialized with shape: {self.weights.shape}")
 
-    async def try_sync_metagraph(self, ttl: int) -> bool:
-        def sync_metagraph(endpoint):
-            # Update self.metagraph
-            self.metagraph = bt.subtensor(endpoint).metagraph(self.config.netuid)
+        self.uids_to_eval: typing.Dict[str, typing.List] = {}
+        bt.logging.info("Initializing score database")
+        self.score_db = ScoreDB("scores.db")
+        bt.logging.info("Score database initialized")
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:125]: Ignore, run from project root.

File: neurons/validator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py]: nope, fixed.

File: FLockDataset/constants.py [hunk 8]
```diff
@@ -1,40 +1,34 @@
...
+        """Create a ChainCommitment from a dictionary"""
+        if not data:
+            return None
 
-# validator weight moving average term. alpha = 1-lr.
-lr = 0.2
+        try:
+            return cls(id=str(data.get("id", 0)), repo=data.get("repo", 0), bench=data.get("bench", 0))
+        except Exception:
+            return None
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/constants.py:30]: ez fix.

File: tests/FlockDataset/utils/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+def node():
+    """Create a mock subtensor with local node connection"""
+    return bt.subtensor("test")
+
+
+def test_read_chain_commitment(node):
+    """Test reading commitment data from another neuron on the chain"""
+    subnet_uid = 257
+
+    key = "5FZGwrY9Ycz8m6fq5rpZCgoSrQWddb7SnZCr3rFU61auctG2"
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply tests/FlockDataset/utils/test_utils.py:16]: fixed.

File: FLockDataset/utils/chain.py [hunk 10]
```diff
@@ -18,4 +21,102 @@ def assert_registered(wallet: bt.wallet, metagraph: bt.metagraph) -> int:
...
+    try:
+        # Convert dict to JSON string
+        json_str = json.dumps(data)
+
+        # Pass the string directly - let bittensor handle the encoding
+        result = node.commit(wallet, subnet_uid, json_str)
+        return result
+    except Exception as e:
+        print(f"Failed to write chain commitment: {str(e)}")
+        return False
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/utils/chain.py:48]: fixed.

File: FLockDataset/validator/validator_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/validator/validator_utils.py:14]: Fixed.

File: neurons/validator.py [hunk 11]
```diff
@@ -16,204 +16,338 @@
...
+            default="~/data/hf_cache",
+            help="Directory to store downloaded model files.",
+        )
+
+        parser.add_argument(
+            "--data_dir",
+            type=str,
+            default="~/data/training_data",
+            help="Directory to store miner datasets.",
+        )
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:71]: add __init checks.

File: neurons/validator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:86]: done.

File: tests/FlockDataset/validators/test_evaluation.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+def test_evaluation():
+    loss = 9999999999999999
+    benchmark_loss = 0.1
+    expected_score = 0.0
+
+    score = compute_score(loss, benchmark_loss)
+    assert np.isclose(
+        score, expected_score, rtol=1e-9
+    ), f"Expected score: {expected_score}, but got: {score}"
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply tests/FlockDataset/validators/test_evaluation.py:13]: added coverage.

File: FLockDataset/validator/validator_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/validator/validator_utils.py:14]: added guards.

File: neurons/validator.py [hunk 13]
```diff
@@ -16,204 +16,340 @@
...
+            bt.logging.warning("Failed to sync metagraph")
+            return
 
-    async def run_step(self):
+        bt.logging.info("Getting current UIDs and hotkeys")
+        current_uids = self.metagraph.uids.tolist()
+        hotkeys = self.metagraph.hotkeys
+        bt.logging.info(f"Current UIDs: {current_uids}")
+
+        base_score = 1.0 / 255.0  # Default initial weight
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:154]: resolved.

File: neurons/validator.py [hunk 14]
```diff
@@ -16,204 +16,340 @@
...
+        bt.logging.info("Normalizing scores")
+        normalized_scores = {}
+        for uid in uids_to_eval:
+            if scores_per_uid[uid] is not None and scores_per_uid[uid] != 0:
+                bt.logging.debug(
+                    f"Computing normalized score for UID {uid} with raw score {scores_per_uid[uid]}"
+                )
+                normalized_score = compute_score(
+                    scores_per_uid[uid], competition.bench
+                )
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:318]: added check.

File: FLockDataset/utils/chain.py [hunk 15]
```diff
@@ -18,4 +21,101 @@ def assert_registered(wallet: bt.wallet, metagraph: bt.metagraph) -> int:
...
+                parts.append(bytes(seg))
+            elif isinstance(seg, (list, tuple)):
+                # if seg[0] is also list/tuple, dive in one level
+                if seg and isinstance(seg[0], (list, tuple)):
+                    for inner in seg:
+                        parts.append(bytes(inner))
+                else:
+                    parts.append(bytes(seg))
+            else:
+                print(f"  ⚠️ unexpected segment #{idx} type={type(seg)}:", seg)
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/utils/chain.py:96]: added logging.

File: FLockDataset/utils/chain.py [hunk 16]
```diff
@@ -18,4 +21,101 @@ def assert_registered(wallet: bt.wallet, metagraph: bt.metagraph) -> int:
...
+            return None
+
+        comp = Competition.from_dict(data)
+        print("🔍 Competition.from_dict →", comp)
+        return comp
 
+    except Exception as e:
+        print("Unhandled exception in read_chain_commitment:", e)
+        import traceback; traceback.print_exc()
+        return None
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/utils/chain.py:121]: resolved.

File: neurons/miner.py [hunk 17]
```diff
@@ -46,17 +52,32 @@ async def main(config: bt.config):
...
 
-    model_id_with_commit = ModelId(namespace=config.hf_repo_id, commit=commit_id,
-                                   competition_id=constants.ORIGINAL_COMPETITION_ID)
-    bt.logging.success(f"Now committing to the chain with model_id: {model_id_with_commit}")
-    print(f"Now committing to the chain with model_id: {model_id_with_commit}")
+    competition: Optional[Competition] = read_chain_commitment(
+        constants.SUBNET_OWNER_HOTKEY, subtensor, config.netuid
+    )
+    if competition is None:
+        bt.logging.error("Failed to read competition commitment")
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/miner.py]: casted.

File: FLockDataset/validator/trainer.py [hunk 18]
```diff
@@ -23,46 +31,76 @@ class LoraTrainingArguments:
...
-def train_lora(lucky_num: int) -> float:
     # set the same random seed to detect duplicate data sets
+    from dotenv import load_dotenv
+
+    load_dotenv()
+    os.environ["PYTHONHASHSEED"] = str(lucky_num)
 
     torch.manual_seed(lucky_num)
     torch.cuda.manual_seed(lucky_num)
     torch.cuda.manual_seed_all(lucky_num)
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply FLockDataset/validator/trainer.py:88]: added.

File: neurons/validator.py [hunk 19]
```diff
@@ -16,204 +16,343 @@
...
         )
         parser.add_argument(
             "--miner_sample_size",
             type=int,
-            default=3, 
+            default=3,
             help="Number of miners to sample for each block.",
         )
+        parser.add_argument("--netuid", type=int, help="The subnet UID.")
+
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply neurons/validator.py:58]: added.

File: tests/FlockDataset/validators/test_scoring.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    ), f"Expected score: {expected_score}, but got: {score}"
+
+def test_none_loss_evaluation():
+    loss = None
+    benchmark_loss = 0.1
+    expected_score = 1.0 / constants.NUM_UIDS 
+    score = compute_score(loss, benchmark_loss)
+    assert np.isclose(
+        score, expected_score, rtol=1e-9
+    ), f"Expected score: {expected_score}, but got: {score}"
```
[the remark this answers is not in the document]
Comment (0xcacti) [reply tests/FlockDataset/validators/test_scoring.py]: fixed.
